### ISSUE TITLE
Implement source maps support to locate JavaScript errors in code blocks

### ISSUE THREAD

**Issue by tslocke** [pegjs/pegjs#93]
For example:

```
$ pegjs syntax.pegjs 
Unexpected token |
```

**Comment by fresheneesz** [comment]
This seems like one of the most important things a parser should do right? If you don't know where your syntax error is, a parser is pretty unusable for development. I'm having a really hard time figuring out where in my pegjs file my parser is failing.

**Comment by StoneCypher** [comment]
👍

**Comment by StoneCypher** [comment]
I hit this for the first time yesterday.  It was _**rough**_ sorting out how to fix it, and the error I got, while correct, was deeply counter-intuitive.

I had intended to write `{ term: 'whatever', value, location: location() }` as the rule.  Instead, I wrote `{ term: 'whatever', value, location() }` (the `location` key is missing,) and received as an error `Unexpected token }`.

Strictly speaking that's correct, but I went off on a fruitless search for aberrant close-braces. I ended up finding it with a differ ☹️ 

With a line number, this would have been trivial.  Pls halp ❤️

**Comment by futagoza** [comment]
I'll look into this after 0.11, but as I haven't done much with source maps before (which is the only way to track error's like this), I can't guarantee a date or version. In the meantime, if anyone is inclined to, they're welcome to send a PR for this 😄

**Comment by michael-brade** [comment]
Yes! I just started working with source maps since they are broken in LiveScript. And now I'm missing them in pegjs :smile:

A popular lib for implementation: https://github.com/mozilla/source-map and a tutorial: https://hacks.mozilla.org/2013/05/compiling-to-javascript-and-debugging-with-source-maps/

Here is a nice tool to help during development: http://sokra.github.io/source-map-visualization/
And for quick usage: https://github.com/lydell/source-map-visualize

**Comment by deostroll** [comment]
:+1: 

Would want source maps not just for stacktrace, but also debugging in, say, vs code or chrome dev tools...since I write my initialization logic separately (in another file) and concat them before parser build.

**Comment by StoneCypher** [comment]
I like the idea of sourcemaps.

While we wait, if you actually need them, there's an easy way to get them: parse your handlers in something that does sourcemaps, like `typescript`, then staple it to the front with string editing as if it was a `peg preamble`, then do a tiny amount of offset math to find the corrected location

That's super ugly and hacky, but it's also durable and something you can do in an hour until this is done properly

Also, I'm looking at my 2017 comment, and I realize in retrospect that for what I was doing, there was a much easier solution: the handlers should have been enforcing the things I was trying to trace with exceptions.  I was just not javascripting effectively at that time.

I have no idea how sourcemaps actually work, so I have no opinions on proper implementation.

dmajda was wrong to cite `i don't want your code` at @Mingun .  That is the same thing Mingun is doing to me, now, in trying to resurrect the library.

The library has needs, the userbase has needs, and we're more interested in our personal intent than solving those needs.  

That's why in 2020 the `es()` function that allows use by packagers, compilers, and transcompilers, which was added in 2017, still hasn't been released.

Because two maintainers "didn't want the code."

It's time for a maintainer that does want the code
